### ISSUE TITLE
Try-It with OAUTH2 Auth Code not passing Authorization header

### ISSUE THREAD

**Issue by CameronGo** [mulesoft/api-console#291]
When I use Try-It with an OAUTH2 Auth Code, I am presented with my authorization end point, enter credentials and then I sit waiting with no response.  Investigating on my implementation, I can see that there are no Authorization headers being received by the API implementation.  See the attached below examples.  Am I missing something with how this is intended to function or can I simply not use Try-It with this type of token?

These are the headers from the request when I make it from the api-console:
```json

{  
    "Host":"api-cam.domain.com",
    "X-Real-IP":"123.123.123.123",
    "X-Forwarded-Proto":"https",
    "Connection":"close",
    "Access-Control-Request-Method":"GET",
    "Origin":"https:\\/\\/chlorine:8446",
    "User-Agent":"Mozilla\\/5.0 (Windows NT 10.0; WOW64) AppleWebKit\\/537.36 (KHTML, like Gecko) Chrome\\/51.0.2704.84 Safari\\/537.36",
    "Access-Control-Request-Headers":"authorization",
    "Accept":"*\\/*",
    "Referer":"https:\\/\\/chlorine:8446\\/api-console\\/dist\\/?raml=\\/pathtoraml\\/api.raml",
    "Accept-Encoding":"gzip, deflate, sdch, br",
    "Accept-Language":"en-US,en;q=0.8"
}
```

These are the headers from the request when I make it from POSTMAN:

```json
{  
    "Host":"api-cam.domain.com",
    "X-Real-IP":"123.123.123.123",
    "X-Forwarded-Proto":"https",
    "Connection":"close",
    "Authorization":"Bearer 539e64d5774624583839fd03ad4e4160e63be73c",
    "Cache-Control":"no-cache",
    "User-Agent":"Mozilla\\/5.0 (Windows NT 10.0; WOW64) AppleWebKit\\/537.36 (KHTML, like Gecko) Chrome\\/51.0.2704.84 Safari\\/537.36",
    "Postman-Token":"a9e61bd7-db16-c001-bbd0-977261feb507",
    "Accept":"*\\/*",
    "Accept-Encoding":"gzip, deflate, sdch, br",
    "Accept-Language":"en-US,en;q=0.8"
}
```

**Comment by CameronGo** [comment]
After more troubleshooting, this definitely seems to be related to CORS.  If I enable CORS on my API and use a RewriteRule to get the OPTIONS then I am able to get the Auto Code working.  I don't love this option; however, and am interested in the proxy functionality I've seen mentioned, but the documentation only barely touches on the topic and I'm not sure exactly how it fits into the solution.

Can anyone provide more information on how to use the proxy settings effectively?

For anyone's reference who needs it, here are the exact settings I put in my .htaccess to get CORS working on the API on my apache2 server in test.  Also note; however, that I had to use something similar on my oauth2 provider to get client credentials auth working, which then broke the auth code I'm using here, so I need to sift through that and figure out how the two are related.

```
Header always set Access-Control-Allow-Origin "*"
Header always set Access-Control-Max-Age "1000"
Header always set Access-Control-Allow-Headers "X-Requested-With, Content-Type, Origin, Authorization, Accept, Client-Security-Token, Accept-Encoding"
Header always set Access-Control-Allow-Methods "POST, GET, OPTIONS, DELETE, PUT"
RewriteCond %{REQUEST_METHOD} OPTIONS
RewriteRule ^(.*)$ $1 [R=200,L]
```

**Comment by CameronGo** [comment]
After a lot of tinkering and educating myself on CORS, the solution is working as expected.  For anyone else who comes along looking for similar input, I had to enable CORS and enable a rewite rule in my .htaccess file on both my OAUTH2 and API endpoint.  The issue I encountered after my enabling CORS was due to the fact I'd enabled the Access-Control-Allow-Origin header in both my apache2 conf file and my .htaccess file, which then returns two values for the Origin header and is not allowed.

For those who aren't familiar with CORS, as was my case, the Header always set ... affectively allows JavaScript from the api-console, that is running in your browser to query data from a URL host (Origin) that is different from that of the api-console.  In our case our api-portal is running on a host (dev.domain.com) so when that JS tries to communicate with oauth2.domain.com or api.domain.com those respective web servers do not allow the request.  Note: We did not implement the wildcard * for our Origin, but rather used a regex statement in the .htaccess that will effectively allow the use of multiple possible Origin values.

The latter issue with the RewriteRule is to address "Preflighted" requests.  There is a lot written up on these and how they work, so I won't get into that, but suffice to say the excerpt from the .htaccess file above worked to address this.

The last comment I'll add is regarding the use of a proxy.  I saw that mentioned in the documentation, but with no real context I couldn't determine how or why this would be used.  The answer is that if you implement both your api-console and your API under the same hostname (and in my case would have to include OAUTH2 provider) then they will all be from the same Origin.  In this case you do not need to enable CORS b/c the JavaScript requests from the one site will already be allowed to the others.  Our reverse proxy sitting in front of our APIs is running on nginx, so we could have deployed our api-console behind that as well, but have opted not to and so must have CORS enabled on the API.

Hopefully this gives a comprehensive answer to anyone looking into this in the future even if it does cover subjects that could be independently researched elsewhere.
